### PR TITLE
Add a 1000 limit on the number of hostnames discovered via mdns

### LINKED ISSUE #886: High CPU usage on USG when using Home Assistant
**Context**
 - CLI Version: 1.41.0
 - Platform: USGRouter 3.10.107-UBNT

**Description**
I have recently noticed that the CPU usage on my USG would rapidly increase, to the point where I start having network issues unless I restart NextDNS on the USG (the memory usage does not increase noticeably though). The problem appeared when I started running Home Assistant (I am running Core version 2023.11.2). I suspect this is linked to Home Assistant also because, if I run `nextdns discovered`, the output is flooded by `homeassistant???????.local.` records. Roughly one hour after a `nextdns restart`, **I found 17,760 such records**:

```
"homeassistant???????.local.": [
            "192.168.???.???"
        ],
```

This only happens on the subnet I use for IoT devices, the other IP addresses that Home Assistant uses do not appear, so it could be linked to some particularly chatty integration.

**Config**
```
admin@USGRouter:~$ nextdns config
debug false
control /var/run/nextdns.sock
cache-max-age 0s
mdns all
timeout 5s
setup-router true
log-queries false
cache-size 10MB
max-ttl 5s
report-client-info true
discovery-dns 
detect-captive-portals false
use-hosts true
listen localhost:53
profile 192.168.X.Y/24=??????
profile ??????
hardened-privacy false
bogus-priv true
max-inflight-requests 256
auto-activate true
admin@USGRouter:~$ 
```

I am using two different NextDNS profiles (which is the only config I customized), but I already tried to remove one, and it did not make any difference about this problem. I have not seen anything abnormal in NextDNS logs.

For now I will uninstall NextDNS from my USG, but I would be happy to do more investigation if that is useful.

### PR DESCRIPTION
**Pull request by rs**
Fixes #886

### CODE REVIEW

File: discovery/mdns.go [hunk 1]
```diff
@@ -215,21 +222,42 @@ func (r *MDNS) read(ctx context.Context, conn *net.UDPConn) {
...
 				h := []byte(name)
 				lowerASCIIBytes(h)
 				key := absDomainName(h)
-				r.addrs[addr] = appendUniq(r.addrs[addr], name)
-				r.names[key] = appendUniq(r.names[key], addr)
+				addEntry(r.addrs, addr, name)
+				addEntry(r.names, key, addr)
+				for len(r.addrs) > mdnsMaxEntries {
+					r.removeOldestEntry()
+				}
```
Comment (justfalter) [inline discovery/mdns.go:236]: @rs The issue still seems to exist when testing using the build you gave me in this comment: https://github.com/nextdns/nextdns/issues/886#issuecomment-1890922498

At present, nextdns is reporting over 5,000 discovered mdns hosts:

```
# nextdns discovered|jq '.mdns | length'
5081
```

The name of my HA install is `hass`,  and here are just a sample of the mdns entries:

```
  "hass9956.local.": [
    "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0",
    "fe80::fefa:6a68:d08f:4e20"
  ],
  "hass9965.local.": [
    "fe80::1c9a:80a1:12dc:e39f",
    "10.200.0.3"
  ],
  "hass9967.local.": [
    "fe80::1c9a:80a1:12dc:e39f",
    "10.200.0.3"
  ],
  "hass9971.local.": [
    "fe80::fefa:6a68:d08f:4e20",
    "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0",
    "192.168.55.3"
  ],
```


Looking at the code I'm commenting on, I believe that you should be checking the length of `r.names`. In my environment, the number of addresses seems to be stable at 66 entries:

```
# nextdns discovered|jq '.mdns | .[] | flatten[]'|sort | uniq -c|wc -l
66
```

However, I do see that the HA addresses are associated with a large number of mdns hostnames:
```
# nextdns discovered|jq '.mdns | .[] | flatten[]'|sort | uniq -c|sort -n|tail -n 6
      3 "10.200.197.10"
   2379 "10.200.0.3"
   2419 "192.168.55.3"
   3678 "fe80::1c9a:80a1:12dc:e39f"
   3859 "fd23:62ab:6fcf:b69b:fe98:6ac0:12cd:bde0"
   3860 "fe80::fefa:6a68:d08f:4e20"
```

**Comment by rs** [comment]
Please try again

**Comment by rs** [comment]
@justfalter would you be able to test again with the last snapshot?

**Comment by justfalter** [comment]
> @justfalter would you be able to test again with the last snapshot?

Sorry for the delay. I'm testing `nextdns version 0.0.0-SNAPSHOT-b4b9696`, and it does seem to limit things to 1000 mdns entries, which solves the memory issue.

The CPU utilization varies between 4% and 12% on my udm pro se. Each call to [removeOldestEntry](https://github.com/nextdns/nextdns/pull/903/files#diff-b7ec724c43603415c8a75c6b78c158a3b541ad6e8752508d15ae46330bb057e0R243-R260) means iterating over all 1000 entries in the map.  The passing around of `mdnsEntry` by value means a lot of value copying is going on.

Two suggestions:
1. Use `*mdnsEntry` to eliminate copying values.
2. Introduce a data structure that would give you O(1) access to the oldest entry. A circular buffer would work -- just has to be a slice of `mdnsMaxEntries` in length. You would know what entry to prune when you go to overwrite it.... you wouldn't even need to keep track of when the entry was added.

**Comment by justfalter** [comment]
> Introduce a data structure that would give you O(1) access to the oldest entry. A circular buffer would work -- just has to be a slice of mdnsMaxEntries in length. You would know what entry to prune when you go to overwrite it.... you wouldn't even need to keep track of when the entry was added.

Thinking about it more, in order to support "updating" an existing entry, a ring-buffer using a slice wouldn't cut it. .. Maybe a linked-list?

**Comment by justfalter** [comment]
Checking in with the process, again, and CPU usage is now sitting between 12 and 20%. Memory usage seems to be stable.

**Comment by rs** [comment]
There is actually no copy happening and the problem is not the search of the oldest name but the unique append of the name in the addrs list. I pushed an optimized version of this function, let me know if it's better.

**Comment by justfalter** [comment]
With d82c635, I consistently see between 3 to 3.7% CPU, occasionally bouncing up to 8%.  When I enable the firewall rules that drop MDNS traffic from HA, CPU usage drops to 0%, occasionally bouncing up to 2%. 

It'd be really convenient if the nextdns cli were wired up with pprof... You could wire it up so that SIGUSR1 would would start a 5 or 10 second profile, so we could see where time was being spent.

**Comment by rs** [comment]
Agreed on pprof. I will merge this PR as I feel it’s good enough for now. This is still an edge case that should probably be fixed at the source.

**Pull request merged**